soc/silabs/common/soc_power_pmgr.c

```diff
@@ -46,31 +47,25 @@ void pm_state_set(enum pm_state state, uint8_t substate_id)
 		break;
 	}
 
-	/* FIXME: When this function is entered the Kernel has disabled
-	 * interrupts using BASEPRI register. This is incorrect as it prevents
-	 * waking up from any interrupt which priority is not 0. Work around the
-	 * issue and disable interrupts using PRIMASK register as recommended
-	 * by ARM.
-	 */
-
-	/* Set PRIMASK */
-	__disable_irq();
-	/* Set BASEPRI to 0 */
-	irq_unlock(0);
-
 	LOG_DBG("Entry to energy mode %d", energy_mode);
 
-	if (energy_mode != SL_POWER_MANAGER_EM0) {
+	if (energy_mode == SL_POWER_MANAGER_EM4) {
+		sl_power_manager_enter_em4();
+	} else if (energy_mode != SL_POWER_MANAGER_EM0) {
+		/* Calling the tracing and hook functions provided in arch_cpu_idle(). */
+		if (IS_ENABLED(CONFIG_TRACING)) {
+			sys_trace_idle();
+		}
+		if (IS_ENABLED(CONFIG_ARM_ON_ENTER_CPU_IDLE_PREPARE_HOOK)) {
+			z_arm_on_enter_cpu_idle_prepare();
```

Comment:
`IS_ENABLED` doesn't work here, since the symbol definition is behind `#if defined(CONFIG_ARM_ON_ENTER_CPU_IDLE_PREPARE_HOOK)`:

```
zephyr/soc/silabs/common/soc_power_pmgr.c:60:25: warning: implicit declaration of function 'z_arm_on_enter_cpu_idle_prepare' [-Wimplicit-function-declaration]
   60 |                         z_arm_on_enter_cpu_idle_prepare();
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```